locator/token_metadata.cc

```diff
@@ -773,122 +710,107 @@ token_metadata_impl::has_pending_ranges(sstring keyspace_name, inet_address endp
-    for (auto& x : bootstrap_addresses) {
-        auto& addr = x.first;
-        auto& t = x.second;
-        tmp[addr].insert(t);
-    }
-    for (auto& x : tmp) {
-        auto& endpoint = x.first;
-        auto& tokens = x.second;
-        all_left_metadata->update_topology(endpoint, get_dc_rack(endpoint), node::state::joining);
-        all_left_metadata->update_normal_tokens(tokens, endpoint).get();
-        auto address_ranges = strategy.get_ranges(endpoint, *all_left_metadata).get0();
-        for (const dht::token_range& x : address_ranges) {
-            new_pending_ranges.emplace(x, endpoint);
-        }
-        all_left_metadata->_impl->remove_endpoint(endpoint);
-    }
-    all_left_metadata->_impl->sort_tokens();
-}
-
 future<> token_metadata_impl::update_pending_ranges(
         const token_metadata& unpimplified_this,
         const abstract_replication_strategy& strategy, const sstring& keyspace_name, dc_rack_fn& get_dc_rack) {
     tlogger.debug("calculate_pending_ranges: keyspace_name={}, bootstrap_tokens={}, leaving nodes={}, replacing_endpoints={}",
         keyspace_name, _bootstrap_tokens, _leaving_endpoints, _replacing_endpoints);
     if (_bootstrap_tokens.empty() && _leaving_endpoints.empty() && _replacing_endpoints.empty()) {
         tlogger.debug("No bootstrapping, leaving nodes, replacing nodes -> empty pending ranges for {}", keyspace_name);
-        set_pending_ranges(keyspace_name, std::unordered_multimap<range<token>, inet_address>(), can_yield::no);
+        _pending_ranges_interval_map.erase(keyspace_name);
         return make_ready_future<>();
     }
 
     return async([this, &unpimplified_this, &strategy, keyspace_name, &get_dc_rack] () mutable {
-        std::unordered_multimap<range<token>, inet_address> new_pending_ranges;
-        calculate_pending_ranges_for_replacing(unpimplified_this, strategy, new_pending_ranges);
-        // Copy of metadata reflecting the situation after all leave operations are finished.
-        auto all_left_metadata = make_token_metadata_ptr(clone_after_all_left().get0());
-        calculate_pending_ranges_for_leaving(unpimplified_this, strategy, new_pending_ranges, all_left_metadata);
-        // At this stage newPendingRanges has been updated according to leave operations. We can
-        // now continue the calculation by checking bootstrapping nodes.
-        calculate_pending_ranges_for_bootstrap(strategy, new_pending_ranges, all_left_metadata, get_dc_rack);
-        all_left_metadata->clear_gently().get();
-
-        // At this stage newPendingRanges has been updated according to leaving and bootstrapping nodes.
-        set_pending_ranges(keyspace_name, std::move(new_pending_ranges), can_yield::yes);
+        // new_token_metadata incorporates all the changes from leaving, bootstrapping and replacing
+        const auto new_token_metadata = token_metadata([&]() -> std::unique_ptr<token_metadata_impl> {
```

Comment:
Style: I prefer immediately-invoked lambdas to be called with std::invoke rather than the function call operator. The reason is that introducing the lambda creates the expectation that it isn't immediately called (as that's how most lambdas are used), and then the expectation is shattered when the call operator is seen. Usually one has to re-read the code with the new knowledge. With std::invoke() you read the code knowing it's going to run now.

No need to change for now, just keep in mind for the future.